keps/sig-apps/3140-TimeZone-support-in-CronJob/README.md

```diff
@@ -0,0 +1,509 @@
+specifying the name of the time zone to be used. Missing or empty value of the
+field indicates the current behavior, which relies on the time zone of the
+kube-controller-manager process.
+
+In the API code, that looks like:
+
+```golang
+
+type CronJobSpec struct {
+
+    // The schedule in Cron format, see https://en.wikipedia.org/wiki/Cron.
+    Schedule string
+
+    // Time zone for the above schedule
+    TimeZone *string
+
+}
+```
+
+The value provided in `TimeZone` field will be validated against the embedded
+golang timezone database.
+
+### CronJob controller
+
+CronJob controller will use non-nil, non-empty value from `TimeZone` field when
+parsing the schedule and during scheduling the next run time. Additionally, the
+time zone will be reflected in the `.status.lastSuccessfulTime` and `.status.lastScheduleTime`.
+In all other cases the controller will maintain the current behavior.
+
+### Test Plan
+
+Unit and integration tests covering the time zone mechanics of CronJob, including:
+
+- defaulting
+- validation
+- creating CronJob
+- updating CronJob
+
+Additionally, all of tests will be performed with feature gate enabled and disabled.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Functionality implemented behind feature gate:
+  - TimeZone field added to API (kube-apiserver)
+  - CronJob controller reacting to new field (kube-controller-manager)
+
+#### Beta
+
+TBD
+
+#### GA
+
+TBD
+
+<!---
+#### Beta
+
+- Gather feedback from developers and surveys
+- Complete features A, B, C
+- Additional tests are in Testgrid and linked in KEP
+
+#### GA
+
+- N examples of real-world usage
+- N installs
+- More rigorous forms of testing—e.g., downgrade tests and scalability tests
+- Allowing time for feedback
+
+**Note:** Generally we also wait at least two releases between beta and
+GA/stable, because there's no opportunity for user feedback, or even bug reports,
+in back-to-back releases.
+
+**For non-optional features moving to GA, the graduation criteria must include
+[conformance tests].**
+
+[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md
+
+#### Deprecation
+
+- Announce deprecation and support policy of the existing flag
+- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
+- Address feedback on usage/changed behavior, provided on GitHub issues
+- Deprecate the flag
+-->
+
+### Upgrade / Downgrade Strategy
+
+- Upgrades
+
+When upgrading from a release without this feature to a release with `TimeZone`
+users should not notice any change in behavior. The default value for `TimeZone`
+is nil, which is equal to current behavior for backwards compatibility.
+
+- Downgrades
+
+When downgrading from a release with this feature, to a release without `TimeZone`
+there are a few cases:
+  1. If TimeZone feature gate was enabled and user specified a TimeZone, the newly
```

Comment:
How will you test to be sure the cronjob controller correctly handles going from timezone aware to non-timezone aware?

Reply:
Manual testing of a cronjob with time zone specified, I can't think of a better approach.

Reply:
I've mentioned that in the test plan, that we want to enabled with the feature on and off to ensure the controller behaves as expected. As much as possible those tests will be automated in integration, but not everything is possible there.